File: algorithms/quicksort.js
function swap(list, i, j) {
  var tmp = list[i];
  list[i] = list[j];
  list[j] = tmp;
}
function partition(list, low, high) {
  var mid = Math.floor((low + high)/2);
  swap(list, mid, high);
  var pivot = high;
  high--;
  while (low <= high) {
    if (list[low] >= list[pivot]) {
      swap(list, low, high);
      high--;
    } else {
      low++;
    }
  }
  swap(list, low, pivot)
  return low;
}
function quicksort(list, low, high) {
  if (low >= high) return;
  var pivot = partition(list, low, high);
  quicksort(list, low, pivot - 1);
  quicksort(list, pivot + 1, high);
}

var l = [5, 2, 1, 6, 7, 3, 2, 6, 1, 9, 8, 0, 100, 1, 35];
quicksort(l, 0, l.length - 1);
console.log(l);

File: algorithms/sorting/quicksort.c
#include <stdio.h>
#include <malloc.h>
void printArray(int *array, int low, int high) {
	while(low <= high) {
		printf("%d ", array[low++]);
	}
	printf("\n");
}
void swap(int *a, int *b) {
	int tmp = *a;
	*a = *b;
	*b = tmp;
}
/*
 *  * pick a pivot index between low and high inclusive
 *   */
int getPivot(int *array, int low, int high) {
	return low;
}
/*
 *  * @return the position of the pivot
 *   */
int partition(int *array, int low, int high, int pivot) {
	int cur = low;
	swap(&array[pivot], &array[high]);
	 
	while (low <= high) {
		if(array[low] < array[high]) {
			swap(&array[low], &array[cur]);
			cur++;
		}
		low++;
	}
	swap(&array[cur], &array[high]);
	return cur;
}
 
void qsortLH(int *array, int low, int high) {
	if(low < high) {
		int pivot = getPivot(array, low, high);
		 
		pivot = partition(array, low, high, pivot);
		qsortLH(array, low, pivot - 1);
		qsortLH(array, pivot + 1, high);
	}
}
 
int main(int argc, char**argv) {
	int *array = (int *) malloc(sizeof(int)*8);
	array[0] = 1;
	array[1] = 4;
	array[2] = 6;
	array[3] = 58;
	array[4] = 4;
	array[5] = 2;
	array[6] = 7;
	array[7] = 23;
	printArray(array, 0, 7);
	qsortLH(array, 0, 7);
	printArray(array, 0, 7);
	return 0;
}
